readme.md
# http_errors

[![tag](https://img.shields.io/github/tag/ako-deno/http_errors.svg)](https://github.com/ako-deno/http_errors/tags)
![http_errors-ci](https://github.com/ako-deno/http_errors/workflows/http_errors-ci/badge.svg)
[![HitCount](http://hits.dwyl.com/ako-deno/http_errors.svg)](http://hits.dwyl.com/ako-deno/http_errors)

Create HTTP Error for Deno, inspired by [http_errors](https://github.com/jshttp/http_errors).

## API

```js
import { createError, HttpError, IError, Props } from "https://deno.land/x/http_errors/mod.ts";
```

### createError(status: number, message?: string, props?: Props): HttpError
### createError(err: Error, props?: Props): IError
### createError(status: number, props: Props): HttpError
              
Create a new error object with the given status code, message and custom properties.
The error object inherits from `HttpError`.

```js
const err = createError(404, 'Not found LoL!')
```

- `name: NotFoundError` - Error name. 
- `status: 404` - the status code as a number.
- `message` - the message of the error, default to [Deno's status code text](https://deno.land/std/http/http_status.ts).
- `expose: true` - can be used to signal if message should be sent to the client, when `status >= 400 && status < 500` it is true, otherwise false. Can be overwritten by Props.
- `properties` - custom properties to attach to the error object.

### HttpError

HttpError is a abstract class, can't be used to create instance.

```js
new HttpError(404) // Error! Cannot create an instance of an abstract class.
createError(500) instanceof HttpError // true
createError(500) instanceof Error // true
```

## Examples

### Simple Http Error handle

This simple example shows how to use `http_errors` to return a different
respond body based on different HTTP Error.

```js
import {
  serve,
  Response,
} from "https://deno.land/std/http/server.ts";
import { createError, HttpError } from "https://deno.land/x/http_errors/mod.ts";

const server = serve("127.0.0.1:3000");
console.log("Server listening on: 3000");

for await (const req of server) {
  const res: Response = {
    body: "Hello!",
    headers: new Headers(),
  };
  try {
    if (req.url === "/4xx") {
      throw createError(403);
    } else if (req.url === "/5xx") {
      throw Error("DB error!");
    }
  } catch (e) {
    if (e instanceof HttpError && e.expose) {
      res.body = JSON.stringify(e.toJSON());
      res.status = e.status;
      console.warn(e);
    } else {
      const err = createError(500);
      res.body = JSON.stringify(err.toJSON());
      res.status = err.status;
      console.error(e);
    }
    res.headers?.set("Content-Type", "applicatoin/json");
  } finally {
    req.respond(res).catch(() => {});
  }
}

```

You can test this out with the cURL program:

```sh
curl -I -H'Accept: text/html' http://localhost:3000/4xx
```

```sh
curl -I -H'Accept: text/html' http://localhost:3000/5xx
```

# License

[MIT](./LICENSE)


History.md

3.0.0 / 2020-09-17
==================

  * feat: support 1.4.0 unstable mode (#2)
  * breaking: only support >= 1.2.0

2.1.0 / 2020-07-12
==================

  * feat: add "createError(err: Error, props?: Props)" support
  * deps: upgrade std to 0.60.0
  * add hits

2.0.0 / 2020-05-28
==================

  * breaking: Disable create error using HttpError
  * doc: add expose checking before sent to client

1.0.0 / 2020-05-24
==================

  * doc: use deno.land/x url
  * ci: fix ci name
  * init


test/http_errors_test.ts
import {
  assert,
  assertStrictEquals,
  assertThrows,
} from "https://deno.land/std@0.69.0/testing/asserts.ts";
import { Status, STATUS_TEXT } from "../deps.ts";
import { createError, HttpError } from "../mod.ts";

const { test } = Deno;

function randomStr(): string {
  return Math.random().toString(36).substring(2, 15) +
    Math.random().toString(36).substring(2, 15);
}

const STATUS_TEXT_KEYS = Object.keys(Status).filter((k) =>
  typeof Status[k as any] === "number"
);
const STATUS_CODES = STATUS_TEXT_KEYS.map((k) => Status[k as any]);
const ClientServerErrorCodes = STATUS_CODES.filter((code) => {
  const _code = Number(code);
  return _code >= 400 && _code < 600;
});

function isError(err: any): boolean {
  return Object.prototype.toString.call(err) === "[object Error]" ||
    err instanceof Error;
}

for (const code of ClientServerErrorCodes) {
  let _code = Number(code);
  const err = createError(_code);
  const message = STATUS_TEXT.get(_code);
  let name = Status[_code];
  if (!name.endsWith("Error")) {
    name += "Error";
  }

  test(`createError(${_code}) should create error object`, function () {
    assert(isError(err));
  });

  test(`createError(${_code}) should have "message" property of "${message}"`, function () {
    assertStrictEquals(err.message, message);
  });

  test(`createError(${_code}) should have "name" property of "${name}"`, function () {
    assertStrictEquals(err.name, name);
  });

  test(`createError(${_code}) should have "statusCode" property of ${_code}`, function () {
    assertStrictEquals(err.status, _code);
  });

  test(`createError(${_code}) should have "expose" property set properly`, function () {
    if (_code >= 400 && _code < 500) {
      assert(err.expose);
    } else {
      assert(!err.expose);
    }
  });
}

test(`createError(status) when status unknown`, function () {
  assertThrows(
    () => {
      createError(3000);
    },
    TypeError,
    "Unknown HTTP Status Code `3000`",
  );
});

test(`createError(status) when out of HTTP Error`, function () {
  const code = 200;
  assertThrows(
    () => {
      createError(code);
    },
    TypeError,
    `Only 4xx or 5xx status codes allowed, but got \`${code}\``,
  );
});

for (const code of ClientServerErrorCodes) {
  let _code = Number(code);
  const message = randomStr();
  const err = createError(_code, message);
  let name = Status[_code];
  if (!name.endsWith("Error")) {
    name += "Error";
  }

  test(`createError(${_code}, "${message}") should create error object`, function () {
    assert(isError(err));
  });

  test(`createError(${_code}, "${message}") should have "message" property of "${message}"`, function () {
    assertStrictEquals(err.message, message);
  });

  test(`createError(${_code}, "${message}") should have "name" property of "${name}"`, function () {
    assertStrictEquals(err.name, name);
  });

  test(`createError(${_code}, "${message}") should have "statusCode" property of ${_code}`, function () {
    assertStrictEquals(err.status, _code);
  });

  test(`createError(${_code}, "${message}") should have "expose" property set properly`, function () {
    if (_code >= 400 && _code < 500) {
      assert(err.expose);
    } else {
      assert(!err.expose);
    }
  });
}

for (const code of ClientServerErrorCodes) {
  let _code = Number(code);
  const message = randomStr();
  const props = {
    year: 2020,
    state: "quarantine at home",
  };
  const propsStr = JSON.stringify(props);
  const err = createError(_code, message, props);
  let name = Status[_code];
  if (!name.endsWith("Error")) {
    name += "Error";
  }

  test(`createError(${_code}, "${message}", ${propsStr}) should create error object`, function () {
    assert(isError(err));
  });

  test(`createError(${_code}, "${message}", ${propsStr}) should have "message" property of "${message}"`, function () {
    assertStrictEquals(err.message, message);
  });

  test(`createError(${_code}, "${message}", ${propsStr}) should have "name" property of "${name}"`, function () {
    assertStrictEquals(err.name, name);
  });

  test(`createError(${_code}, "${message}", ${propsStr}) should have "statusCode" property of ${_code}`, function () {
    assertStrictEquals(err.status, _code);
  });

  test(`createError(${_code}, "${message}", ${propsStr}) should have "props" property of ${propsStr}`, function () {
    assertStrictEquals(err.year, props.year);
    assertStrictEquals(err.state, props.state);
  });

  test(`createError(${_code}, "${message}", ${propsStr}) should have "expose" property set properly`, function () {
    if (_code >= 400 && _code < 500) {
      assert(err.expose);
    } else {
      assert(!err.expose);
    }
  });
}

for (const code of ClientServerErrorCodes) {
  let _code = Number(code);
  const props = {
    year: 2020,
    state: "quarantine at home",
  };
  const message = STATUS_TEXT.get(_code);
  const propsStr = JSON.stringify(props);
  const err = createError(_code, props);
  let name = Status[_code];
  if (!name.endsWith("Error")) {
    name += "Error";
  }

  test(`createError(${_code}, ${propsStr}) should create error object`, function () {
    assert(isError(err));
  });

  test(`createError(${_code}, ${propsStr}) should have "message" property of "${message}"`, function () {
    assertStrictEquals(err.message, message);
  });

  test(`createError(${_code}, ${propsStr}) should have "name" property of "${name}"`, function () {
    assertStrictEquals(err.name, name);
  });

  test(`createError(${_code}, ${propsStr}) should have "statusCode" property of ${_code}`, function () {
    assertStrictEquals(err.status, _code);
  });

  test(`createError(${_code}, ${propsStr}) should have "props" property of ${propsStr}`, function () {
    assertStrictEquals(err.year, props.year);
    assertStrictEquals(err.state, props.state);
  });

  test(`createError(${_code}, ${propsStr}) should have "expose" property set properly`, function () {
    if (_code >= 400 && _code < 500) {
      assert(err.expose);
    } else {
      assert(!err.expose);
    }
  });
}

test("createError(err)", function () {
  let _err = new Error("LOL");
  (_err as any).status = 404;
  let err = createError(_err);
  assertStrictEquals(err, _err);
  assertStrictEquals(err.name, "Error");
  assertStrictEquals(err.message, "LOL");
  assertStrictEquals(err.status, 404);
  assertStrictEquals(err.statusCode, 404);
  assertStrictEquals(err.expose, true);

  _err = new Error("LOL");
  err = createError(_err);
  assertStrictEquals(err, _err);
  assertStrictEquals(err.name, "Error");
  assertStrictEquals(err.message, "LOL");
  assertStrictEquals(err.status, 500);
  assertStrictEquals(err.statusCode, 500);
  assertStrictEquals(err.expose, false);
});

test("createError(err) with invalid err.status", function () {
  let _err = new Error("Connection refused");
  (_err as any).status = -1;
  let err = createError(_err);
  assertStrictEquals(err, _err);
  assertStrictEquals(err.name, "Error");
  assertStrictEquals(err.message, "Connection refused");
  assertStrictEquals(err.status, 500);
  assertStrictEquals(err.statusCode, 500);
  assertStrictEquals(err.expose, false);
});

test("createError(err, props)", function () {
  let _err = new Error("LOL");
  (_err as any).status = 404;
  let err = createError(_err, {
    id: 1,
  });
  assertStrictEquals(err.name, "Error");
  assertStrictEquals(err.message, "LOL");
  assertStrictEquals(err.status, 404);
  assertStrictEquals(err.statusCode, 404);
  assertStrictEquals(err.id, 1);
  assertStrictEquals(err.expose, true);
});

test("instanceof HttpError", () => {
  assert(createError(500) instanceof HttpError);
});

test("instanceof Error", () => {
  assert(createError(500) instanceof Error);
});


mod.ts
/*!
 * Copyright (c) 2020 Henry Zhuang
 * MIT Licensed
 */

import { Status, STATUS_TEXT } from "./deps.ts";

export abstract class HttpError extends Error {
  name: string;
  message: string;
  status: number;
  statusCode: number;
  expose: boolean = false;
  [key: string]: any
  constructor(code: number, message?: string) {
    super(message);
    if (!Status[code]) {
      throw TypeError(`Unknown HTTP Status Code \`${code}\``);
    }
    if (code < 400 || code >= 600) {
      throw TypeError(
        `Only 4xx or 5xx status codes allowed, but got \`${code}\``,
      );
    }
    if (code >= 400 && code < 500) {
      this.expose = true;
    }

    let className = Status[code];
    if (!className.endsWith("Error")) {
      className += "Error";
    }
    const msg = message != null ? message : STATUS_TEXT.get(code)!;
    this.message = msg;
    this.status = this.statusCode = code;
    this.name = className;
    (Error as any).captureStackTrace(this, this.constructor);
    Object.setPrototypeOf(this, new.target.prototype);
  }

  toString() {
    return `${this.name} [${this.status}]: ${this.message}`;
  }

  toJSON() {
    return {
      name: this.name,
      status: this.status,
      message: this.message,
    };
  }
}

class HttpErrorImpl extends HttpError {}

export interface Props {
  [key: string]: any;
}

export interface IError extends Error {
  status: number;
  statusCode: number;
  expose: boolean;
  [key: string]: any;
}

/**
 * Create a new HttpError.
 *
 * @returns {HttpError}
 * @public
 */

export function createError(
  status: number,
  message?: string,
  props?: Props,
): HttpError;
export function createError(status: number, props: Props): HttpError;
export function createError(err: Error, props?: Props): IError;
export function createError(
  status: any,
  message?: any,
  props?: Props,
): HttpError | Error {
  let err;
  if (status instanceof Error) {
    err = status as IError;
    status = err.status || err.statusCode;

    if (
      typeof status !== "number" ||
      (!Status[status] && (status < 400 || status >= 600))
    ) {
      status = 500;
    }

    props = message;
  } else if (typeof message === "string") {
    err = new HttpErrorImpl(status, message);
    (Error as any).captureStackTrace(err, createError);
  } else {
    props = message;
    err = new HttpErrorImpl(status);
    (Error as any).captureStackTrace(err, createError);
  }

  if (!(err instanceof HttpError) || err.status !== status) {
    // add properties to generic error
    err.expose = status < 500;
    err.status = err.statusCode = status;
  }

  if (props) {
    for (let key in props) {
      if (key !== "status") {
        err[key] = props[key];
      }
    }
  }

  return err;
}


example/server.ts
import {
  serve,
  Response,
} from "https://deno.land/std/http/server.ts";
import { createError, HttpError } from "../mod.ts";

const server = serve("127.0.0.1:3000");
console.log("Server listening on: 3000");

for await (const req of server) {
  const res: Response = {
    body: "Hello!",
    headers: new Headers(),
  };
  try {
    if (req.url === "/4xx") {
      throw createError(403);
    } else if (req.url === "/5xx") {
      throw Error("DB error!");
    }
  } catch (e) {
    if (e instanceof HttpError && e.expose) {
      res.body = JSON.stringify(e.toJSON());
      res.status = e.status;
      console.warn(e);
    } else {
      const err = createError(500);
      res.body = JSON.stringify(err.toJSON());
      res.status = err.status;
      console.error(e);
    }
    res.headers?.set("Content-Type", "applicatoin/json");
  } finally {
    req.respond(res).catch(() => {});
  }
}
